src/main/java/com/mate/dao/jdbcimpl/CarDaoJdbcImpl.java

```diff
@@ -0,0 +1,257 @@
+                transaction.append(", ");
+            }
+        }
+        transaction.append("ON DUPLICATE KEY UPDATE car_id = car_id");
+        String insertQuery = transaction.toString();
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement preparedStatement =
+                        connection.prepareStatement(
+                             insertQuery)) {
+            for (int i = 0; i < drivers.size(); i++) {
+                Driver driver = drivers.get(i);
+                preparedStatement.setLong((i * 2) + 1, carId);
+                preparedStatement.setLong((i * 2) + 2, driver.getId());
+            }
+            preparedStatement.executeUpdate();
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't insert drivers " + drivers, e);
+        }
+    }
+
+    private void deleteAllDriversExceptList(Long carId, List<Driver> exceptions) {
+        int size;
+        if (exceptions == null) {
+            size = 0;
+        } else {
+            size = exceptions.size();
+        }
+        StringBuilder transaction = new StringBuilder(
+                "DELETE FROM cars_drivers WHERE car_id = ? AND NOT driver_id IN (0");
+        for (int i = 0; i < size; i++) {
+            transaction.append(", ?");
+        }
+        transaction.append(");");
+        String insertQuery = transaction.toString();
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement preparedStatement =
+                        connection.prepareStatement(
+                             insertQuery)) {
+            preparedStatement.setLong(1, carId);
+            for (int i = 0; i < size; i++) {
+                Driver driver = exceptions.get(i);
+                preparedStatement.setLong((i) + 2, driver.getId());
+            }
+            preparedStatement.executeUpdate();
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't delete drivers " + exceptions, e);
+        }
+    }
+
+    private List<Driver> getAllDriversByCarId(Long carId) {
+        String selectQuery = "SELECT driver_id as id, name, license_number FROM cars_drivers "
+                + "join drivers d on cars_drivers.driver_id = d.id "
+                + "where car_id = ?";
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement preparedStatement =
+                        connection.prepareStatement(selectQuery)) {
+            preparedStatement.setLong(1, carId);
+            ResultSet resultSet = preparedStatement.executeQuery();
+            List<Driver> drivers = new ArrayList<>();
+            while (resultSet.next()) {
+                drivers.add(parseDriverFromResultSet(resultSet));
+            }
+            return drivers;
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't get all drivers", e);
```

Comment:
I guess we should make exception messages more informative
```suggestion
            throw new DataProcessingException("Can't get all drivers by car id " + carId, e);
```

Reply:
Agree, thanks